Polarity suggestion

blishko opened the pull request:
This adds the possibility for the SAT solver to ask theory solver for suggested polarity once SAT solver has picked new decision variable.

New interface method is provided with implementations for LRA solver and UF solver.
The measured impact in LRA is huge, in UF is small but still positive.

blishko commented:
@aehyvari, I created an option in SMTConfig, which enables this feature by default.
Should this option be also accessible from SMT-LIB scripts? Should a string representation be created?

aehyvari commented on src/tsolvers/lrasolver/LRASolver.C:
```diff
@@ -191,6 +191,32 @@ LRASolver::~LRASolver( )
 
+lbool LRASolver::getPolaritySuggestion(PTRef ptref) const {
+    if (!this->isInformed(ptref)) { return l_Undef; }
+    LVRef var = this->lavarStore.getVarByLeqId(logic.getPterm(ptref).getId());
```
Do we need a method LVRef getVarByPTRef(PTRef tr) const { return getVarByLeqId(logic.getPterm(tr).getId())); }  ?

blishko replied:
> Do we need a method LVRef getVarByPTRef(PTRef tr) const { return getVarByLeqId(logic.getPterm(tr).getId())); } ?

It looks like it could be useful. I can try to add it.

aehyvari commented on src/tsolvers/lrasolver/LRASolver.C:
```diff
@@ -191,6 +191,32 @@ LRASolver::~LRASolver( )
+        negative = true;
+    }
+    // MB: Maybe we will not get any suggestion, if the value is <0|-1/2>, but the bounds are <0|0> and <0|-1>
+//    assert(positive || negative);
```
What's the invariant here?  Can assert(positive || negative) be removed?

blishko replied:
Originally I thought that the invariant would be positive XOR negative (exactly one will be true and the other one false).
But then I ran into a situation where a value of a variable, "x", was <0, -1/2> (-1/2 is the delta), and the constraint was x >= 0. The positive bound on this is a lower bound with value <0,0> and the negative bound is an upper bound with value <0, -1> ( x < 0, i.e, x <= 0 - Delta)
In this case the current value does not satisfy neither positive nor negative bound.
It is kind of a special case but I had to comment out the assert because of this since it is a valid situation.
I just wanted to stress that this assertion is not valid (even though I originally thought it would be) so I left it there as a comment.

Do you suggest to remove it completely?

aehyvari replied:
Hm ok... In that case the 'consistent' suggestion would be to set negative polarity, that is, x < <0,-1>?  But it wouldn't really be consistent in the end because x = <0,-1/2> would break x < <0,-1>.   What does our code do in that case?

I'm not sure what's the right thing to do about the assert.  Perhaps removing it and expanding the comment a little as you do above would be better.

blishko replied:
Yes, neither choice is consistent with the assignment, so the theory solver just says "I don't know" and the decision falls back to the SAT solver.

I will remove the commented assert and add more detailed explanation.

aehyvari commented on src/tsolvers/egraph/EgraphSolver.C:
```diff
@@ -221,6 +221,34 @@ PTRef Egraph::getSuggestion( )
+    PTRef rhs = term[1];
+    assert(enode_store.termToERef.has(lhs) && enode_store.termToERef.has(rhs));
+    if (!enode_store.termToERef.has(lhs) || !enode_store.termToERef.has(rhs)) { return l_Undef; }
+    ERef e_lhs = termToERef(lhs);
```
Looks like the code 236 -- 239 is the same as the beginning of Egraph::unmergeable.  I think the same functionality would be available with less code by just calling bool res = unmergeable(e_lhs, e_rhs, tmp);

blishko replied:
The change you suggest would result in a different behaviour.
With the current implementation, EGraph suggests to make things equal only if they ARE ALREADY in the same equivalence classes. (And suggests to make them not equal if they are unmergeable).

In your scenario, Egraph would suggest to make things equal if it CANNOT prove that they are unmergeable. So it would make suggestions more often.

My original idea was for the Egraph would only suggest polarity if it KNOWS that the opposite polarity would lead to conflict. But I don't mind making it more aggressive suggestions to make things equal. If you think it's better that way I will change it.

aehyvari replied:
I think you're right, just keep it the way you first coded it.

aehyvari commented:
Hi Martin, thanks!  I think that your experiments are so convincingly good
that it might just be better to make it not an option :)

On Tue, Aug 20, 2019 at 4:25 PM Martin Blicha <notifications@github.com>
wrote:

> @aehyvari <https://github.com/aehyvari>, I created an option in
> SMTConfig, which enables this feature by default.
> Should this option be also accessible from SMT-LIB scripts? Should a
> string representation be created?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/usi-verification-and-security/opensmt/pull/32?email_source=notifications&email_token=AB74VMTH5ZOQWIDUO4PBYW3QFP5GTA5CNFSM4INXR5X2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4WPGQY#issuecomment-523039555>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB74VMVLCZWNCJ6DEW2NKXLQFP5GTANCNFSM4INXR5XQ>
> .
>

blishko commented:
I have changed the comments in LRASolver::getPolaritySuggestion and refactored the code of translating PTRef to LVRef a little bit.
Check the two new commits.
